### PR TITLE
KEYCLOAK-2028: Add preemptive access token refresh support

### PR DESCRIPTION
**Pull request by foo4u**
Add a new keycloak.json property and mechanism to automatically
refresh access tokens if they are going to expire in less than a configurable
amount of time.

### CODE REVIEW

File: docbook/auth-server-docs/reference/en/en-US/modules/adapter-config.xml [hunk 1]
```diff
@@ -401,10 +402,23 @@
                 <listitem>
                     <para>
                         The session id is changed by default on a successful login on some platforms to plug a security attack vector (Tomcat 8, Jetty9, Undertow/Wildfly).  Change this to true if you want to turn this off
+                        This is especially useful in when the access token is sent to another client where it could expire before being evaluated.
```
Comment (stianst) [inline docbook/auth-server-docs/reference/en/en-US/modules/adapter-config.xml]: I assume this is a copy/paste error?

**Comment by stianst** [comment]
Looks like there's a compile error. Also, a minor mistake in the docs update.

File: docbook/auth-server-docs/reference/en/en-US/modules/adapter-config.xml (hunk 1 above)
Comment (foo4u) [inline docbook/auth-server-docs/reference/en/en-US/modules/adapter-config.xml]: Merge issue. Will fix.

**Comment by foo4u** [comment]
It complied ok locally. Will retest after updating docs.

**Comment by thomasdarimont** [comment]
@foo4u any chance to rebase this on current master?

**Comment by stianst** [comment]
@mposolda can you review and merge this? Also add a test for it?

**Comment by mposolda** [comment]
Closing as I've created other PR including test https://github.com/keycloak/keycloak/pull/2924

**Pull request closed without merging**